File: clients/go/examples/loadtest_100k.go
package main

import (
	"fmt"
	"log"
	"os"
	"runtime"
	"runtime/pprof"
	"sync"
	"sync/atomic"
	"time"
	blazecache "github.com/blazecache/client"
)

func main() {
	const (
		serverAddr = "127.0.0.1:6792"
		numOps     = 100000
		numWorkers = 100
	)

	fmt.Printf("=== Go Client Load Test: %d operations with %d workers ===\n\n", numOps, numWorkers)

	// Create client
	client, err := blazecache.New(serverAddr)
	if err != nil {
		log.Fatal(err)
	}

	// Test ping first
	if err := client.Ping(); err != nil {
		log.Fatalf("Ping failed: %v", err)
	}
	fmt.Println("✓ Server connection verified")

	var (
		successCount int64
		errorCount   int64
		wg           sync.WaitGroup
		opsPerWorker = numOps / numWorkers
	)

	// Enable CPU profiling
	cpuProfile, err := os.Create("cpu.prof")
	if err != nil {
		log.Fatalf("Failed to create CPU profile: %v", err)
	}
	defer cpuProfile.Close()
	if err := pprof.StartCPUProfile(cpuProfile); err != nil {
		log.Fatalf("Failed to start CPU profile: %v", err)
	}
	defer pprof.StopCPUProfile()

	// Enable memory profiling
	memProfile, err := os.Create("mem.prof")
	if err != nil {
		log.Fatalf("Failed to create memory profile: %v", err)
	}
	defer memProfile.Close()
	defer func() {
		runtime.GC()
		if err := pprof.WriteHeapProfile(memProfile); err != nil {
			log.Fatalf("Failed to write memory profile: %v", err)
		}
	}()

	start := time.Now()

	// Spawn workers
	for w := 0; w < numWorkers; w++ {
		wg.Add(1)
		go func(workerID int) {
			defer wg.Done()
			localSuccess := int64(0)
			localErrors := int64(0)

			for i := 0; i < opsPerWorker; i++ {
				key := fmt.Sprintf("key-%d-%d", workerID, i)
				value := []byte(fmt.Sprintf("value-%d-%d", workerID, i))

				// Set operation
				item := &blazecache.Item{Key: key, Value: value}
				if err := client.Set(item); err != nil {
					localErrors++
					continue
				}

				// Get operation
				result, err := client.Get(key)
				if err != nil || result == nil || string(result.Value) != string(value) {
					localErrors++
					continue
				}

				localSuccess++
			}

			atomic.AddInt64(&successCount, localSuccess)
			atomic.AddInt64(&errorCount, localErrors)
		}(w)
	}

	wg.Wait()
	elapsed := time.Since(start)

	totalOps := atomic.LoadInt64(&successCount) + atomic.LoadInt64(&errorCount)
	success := atomic.LoadInt64(&successCount)
	errors := atomic.LoadInt64(&errorCount)

	fmt.Printf("\n=== Results ===\n")
	fmt.Printf("Total operations: %d\n", totalOps)
	fmt.Printf("Successful: %d (%.2f%%)\n", success, float64(success)/float64(totalOps)*100)
	fmt.Printf("Errors: %d (%.2f%%)\n", errors, float64(errors)/float64(totalOps)*100)
	fmt.Printf("Time elapsed: %v\n", elapsed)
	fmt.Printf("Throughput: %.2f ops/sec\n", float64(success)/elapsed.Seconds())
	fmt.Printf("Avg latency: %.2f µs/op\n", float64(elapsed.Nanoseconds())/float64(success)/1000)
}

File: clients/rust/examples/loadtest_100k.rs
use blazecache_client::{TcpClient, SelectionStrategy};
use std::time::Instant;
use tokio::time::Duration;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let server_addr_str = std::env::var("SERVER_ADDR").unwrap_or_else(|_| "127.0.0.1:6784".to_string());
    // Parse comma-separated server addresses
    let servers: Vec<String> = server_addr_str.split(',').map(|s| s.trim().to_string()).collect();
    let num_ops = std::env::var("NUM_OPS")
        .ok()
        .and_then(|s| s.parse().ok())
        .unwrap_or(100_000);
    let num_workers = std::env::var("NUM_WORKERS")
        .ok()
        .and_then(|s| s.parse().ok())
        .unwrap_or(100);

    println!("=== Rust Client Load Test: {} operations with {} workers ===", num_ops, num_workers);
    println!("Servers: {:?}", servers);

    // Verify server connection
    let test_client = TcpClient::new(servers.clone());
    match test_client.ping().await {
        Ok(_) => println!("✓ Server connection verified\n"),
        Err(e) => {
            eprintln!("✗ Server connection failed: {}", e);
            return Err(e.into());
        }
    }

    let start = Instant::now();
    let mut handles = Vec::new();

    let ops_per_worker = num_ops / num_workers;
    let remainder = num_ops % num_workers;

    for worker_id in 0..num_workers {
        let ops = if worker_id < remainder {
            ops_per_worker + 1
        } else {
            ops_per_worker
        };

        let client = TcpClient::with_strategy(
            servers.clone(),
            SelectionStrategy::RoundRobin,
        );

        let handle = tokio::spawn(async move {
            let mut success = 0;
            let mut errors = 0;

            for i in 0..ops {
                let key = format!("key-{}-{}", worker_id, i);
                let value = format!("value-{}-{}", worker_id, i).into_bytes();

                // SET operation
                match client.set_with_ttl(&key, value.clone(), 3600).await {
                    Ok(_) => success += 1,
                    Err(_) => errors += 1,
                }

                // GET operation
                match client.get(&key).await {
                    Ok(Some(v)) if v == value => success += 1,
                    Ok(_) => errors += 1,
                    Err(_) => errors += 1,
                }
            }

            (success, errors)
        });

        handles.push(handle);
    }

    let mut total_success = 0;
    let mut total_errors = 0;

    for handle in handles {
        let (success, errors) = handle.await?;
        total_success += success;
        total_errors += errors;
    }

    let elapsed = start.elapsed();
    let throughput = (total_success as f64) / elapsed.as_secs_f64();
    let avg_latency = elapsed.as_secs_f64() / (total_success as f64) * 1_000_000.0; // microseconds

    println!("=== Results ===");
    println!("Total operations: {}", total_success + total_errors);
    println!("Successful: {} ({:.2}%)", total_success, (total_success as f64 / (total_success + total_errors) as f64) * 100.0);
    println!("Errors: {} ({:.2}%)", total_errors, (total_errors as f64 / (total_success + total_errors) as f64) * 100.0);
    println!("Time elapsed: {:?}", elapsed);
    println!("Throughput: {:.2} ops/sec", throughput);
    println!("Avg latency: {:.2} µs/op", avg_latency);

    Ok(())
}
